core/src/test/java/io/questdb/test/griffin/engine/window/WindowFunctionUnitTest.java

```diff
@@ -0,0 +1,327 @@
+                    if (rangeLo == Long.MIN_VALUE && rangeHi == 0) {
+                        return new SumDoubleWindowFunctionFactory.SumOverUnboundedRowsFrameFunction(
+                                TestDefaults.createLongFunction(x -> x.getLong(2))
+                        );
+                    }
+                    return new SumDoubleWindowFunctionFactory.SumOverRowsFrameFunction(
+                            TestDefaults.createLongFunction(x -> x.getLong(2)),
+                            rangeLo,
+                            rangeHi,
+                            TestDefaults.createMemoryCARW()
+                    );
+                },
+                Double::sum
+        );
+    }
+
+    @Test
+    public void testFirstOverRowsFuzz() throws Exception {
+        fuzzTestBase(
+                TestUtils.generateRandom(LOG),
+                false,
+                true,
+                rnd -> rnd.nextInt(8) == 0 ? Long.MIN_VALUE : -rnd.nextLong(1024),
+                rnd -> -rnd.nextLong(1024),
+                (rangeLo, rangeHi) -> {
+                    if (rangeLo == Long.MIN_VALUE && rangeHi == 0) {
+                        return new FirstValueDoubleWindowFunctionFactory.FirstValueOverWholeResultSetFunction(TestDefaults.createLongFunction(x -> x.getLong(2)));
+                    }
+                    return new FirstValueDoubleWindowFunctionFactory.FirstValueOverRowsFrameFunction(
+                            TestDefaults.createLongFunction(x -> x.getLong(2)),
+                            rangeLo,
+                            rangeHi,
+                            TestDefaults.createMemoryCARW()
+                    );
+                },
+                (a, b) -> b
+        );
+    }
+
+    @Test
+    public void testAggRangeFuzz() throws Exception {
+        fuzzTestBase(
+                TestUtils.generateRandom(LOG),
+                false,
+                false,
+                rnd -> rnd.nextInt(8) == 0 ? Long.MIN_VALUE : -rnd.nextLong(65536),
+                rnd -> -rnd.nextLong(65536),
+                (rangeLo, rangeHi) -> new SumDoubleWindowFunctionFactory.SumOverRangeFrameFunction(
+                        rangeLo,
+                        rangeHi,
+                        TestDefaults.createLongFunction(x -> x.getLong(2)),
+                        64,
+                        TestDefaults.createMemoryCARW(),
+                        0
+                ),
+                Double::sum
+        );
+    }
+
+    public void fuzzTestBase(
+            Rnd rnd,
+            boolean partitioned,
+            boolean rows,
+            java.util.function.Function<Rnd, Long> rangeLoGen,
+            java.util.function.Function<Rnd, Long> rangeHiGen,
+            java.util.function.BiFunction<Long, Long, BaseDoubleWindowFunction> windowFunctionFactory,
+            java.util.function.BiFunction<Double, Double, Double> sum
+    ) throws Exception {
+        for (int count = 1; count <= 512; count *= 2) {
```

Comment:
count could be something like:

```
final int count = rnd.nextPositiveInt() % 512;
```

outer loop can be removed in favour of the random count. Also, i am not clear why could has to be a power of 2?

Reply:
Yes, removed. 
Power of 2 is not really important - maybe I wanted to test some edge cases related to map reallocation. But randomized values are better, yes.